### ISSUE TITLE
PriorityQueue does not guarantee ordering of elements of equal priority

### ISSUE THREAD

**Issue by eyalroth** [scala/bug#11079]
In the following a code, a priority queue for integers is defined so that there are 3 priorities - one for positive integers, one for zero and one for negative integers.

```scala
scala> def printQ(is: Int*): Unit = {
     |   val q = scala.collection.mutable.PriorityQueue.empty[Int](Ordering.by(_.compareTo(0)))
     |   is.foreach(i => q.enqueue(i))
     |   while(q.nonEmpty){
     |     print(s"${q.dequeue},")
     |   }
     | }

scala> printQ(1,2,3,4)
1,4,3,2,

scala> printQ(1,2,-1,-2,-3,0,3,4)
1,2,4,3,0,-2,-1,-3,
```

The code shows that the ordering of elements with the same priority is arbitrary and is not the same as the insertion order. That is quite counter-intuitive and goes against the definition of a priority queue according to Wikipedia:

> In computer science, a priority queue is an abstract data type which is like a regular queue or stack data structure, but where additionally each element has a "priority" associated with it. In a priority queue, an element with high priority is served before an element with low priority. **If two elements have the same priority, they are served according to their order in the queue.**

I do realize that this behavior is apparent in other programming languages as well. For instance, in Java the javadoc for `java.util.PriorityQueue` fail to mention how it handles elements with the same priority, while `java.util.concurrent.PriorityBlockingQueue` explicitly mentions that the order of elements of equal priority is arbitrary.

I think this is a chance for Scala to make things better. At minimum I would expect the scaladoc to explicitly mention this behavior, and provide a best-practice for how to implement such functionality.

**Comment by diesalbla** [comment]
After looking for a little while on this issue, and through [this StackOverflow answer](https://stackoverflow.com/a/47969819/1002111), I found the following [Python documentation page](https://docs.python.org/3.6/library/heapq.html#priority-queue-implementation-notes).

> - Sort stability: how do you get two tasks with equal priorities to be returned in the order they were originally added? [...]
> 
> A solution [...] is to store entries as 3-element list including the priority, an entry count, and the task.

The StackOverflow answer itself also offers an interesting idea: 
> maybe better define your own `StablePriorityQueue` class, as subclass or wrapper of `PriorityQueue`.

Also, you can just handle those entry numbers outside the `PriorityQueue[A]`, by inserting into  queue tuples with the entry number, and using an `Ordering` that breaks ties with the entry number. 

Given the alternatives, there is little benefit in modifying the `PriorityQueue` to make it stable, as it would cost performance to the benefit of a few use cases. Instead, the behaviour should be  explicitly described in the Scaladoc.

**Comment by eyalroth** [comment]
@diesalbla I am aware of the alternatives. My argument is that the term "priority queue" denotes semantics which are ignored by the default implementation, thus creating confusion amongst Scala developers. I would argue that this confusion had caused / still causing plenty of bugs for developers, some are perhaps not yet discovered since developers might be so certain in their intuitive understanding of the data structure.

I honestly fail to understand how it became that the term "priority queue" got abused to represent a data structure which IMHO strays far from the intuitive understanding of what it should be doing. It is a **queue** after all, which means it dequeues elements in a FIFO order, but it gives priority to certain elements based on an arbitrary priority.

Also, it seems that there's a very simple and still very performant way of implementing it this way. If we take the implementation of a heap, there's no need to store each element as a node in the heap; nodes merely represent a pointer to a queue of elements with equal priority.

In other words, such a data structure should contain a sorted index between a priority to a queue with its elements. I don't see why this implementation couldn't be `O(log(p))` on enqueue (where `p` is the number of different priorities), and `O(1)` on dequeue. It could also allow for concurrency for elements of different priorities (at least on enqueue).

**Comment by NthPortal** [comment]
@eyalroth the [Javadoc for `j.u.PriorityQueue`](https://docs.oracle.com/javase/8/docs/api/java/util/PriorityQueue.html) actually does state how it handles elements with the same priority:

> The head of this queue is the least element with respect to the specified ordering. If multiple elements are tied for least value, the head is one of those elements -- ties are broken arbitrarily.

**Comment by SethTisue** [comment]
> I honestly fail to understand how it became that the term "priority queue" got abused to represent a data structure which IMHO strays far from the intuitive understanding of what it should be doing

also https://en.wikipedia.org/wiki/Priority_queue 's first paragraph says:

> If two elements have the same priority, they are served according to their order in the queue

but then farther down it says

> The Standard Template Library (STL), and the C++ 1998 standard, specifies priority_queue as one of the STL container adaptor class templates. However, it does not specify how two elements with same priority should be served, and indeed, common implementations will not return them according to their order in the queue

**Comment by eyalroth** [comment]
@NthPortal I admit to have missed that. I guess when I read this my mind was interpreting this as "implementation details, I don't care" 😆 

@SethTisue Yep, I've read that, and this is exactly what frustrates me. I don't understand why "common implementations will not return them according to their order in the queue".

**Comment by Jasper-M** [comment]
Ok, I agree. A priority queue is by definition **not** a queue.

**Comment by NthPortal** [comment]
I just checked, and python3's `queue.PriorityQueue` does not respect insertion order on its own. If even python doesn't do it, I'm going to go ahead and say that the Wikipedia article is wrong.

I also think it's worth noting that it's not difficult to wrap your elements so that the queue will respect insertion order if you prefer that and don't care about the performance loss; however, if that was the default, it would be impossible to regain the performance.

**Comment by diesalbla** [comment]
To my view, the current issue comes down to three alternatives for action: 
- Modify the documentation of that class to explicitly indicate that "ties are broken arbitrarily"; or
- Rename the `PriorityQueue` to `Heap`, or `ArrayHeap`, which seems may be more accurate'; or
- Modify the `PriorityQueue` class to ensure that, for tie-breaking, it is also a FIFO queue.

The first one seems straightforward. The second one is easy, but may be confusing to people from other programming languages. It may also cause some compatibility problems. The third one incurs in a performance cost that is to be paid by any program using it.

**Comment by NthPortal** [comment]
I am in favour of the first option there

**Comment by martijnhoekstra** [comment]
As a long time wikipedian, and administrator on the English wikipedia, the use of the text of a wikipedia article as an authoritive source makes me very uneasy

**Comment by SethTisue** [comment]
sadly the word "heap" is also badly overloaded in computer science :-(

**Comment by eyalroth** [comment]
Just to strengthen the confusion argument, take a look at the documentation of `scala.collection.immutable.Queue`:

>  `Queue` objects implement data structures that allow to insert and retrieve elements in a first-in-first-out (FIFO) manner.

Whereas `java.util.Queue` goes:

> Queues typically, but do not necessarily, order elements in a FIFO (first-in-first-out) manner.  Among the exceptions are priority queues, which order elements according to a supplied comparator, or the elements' natural ordering, and LIFO queues (or stacks) which order the elements LIFO (last-in-first-out).

So in fact in Java they made sure to reflect the inconsistency of the terminology in their documentation, which IMHO just shows how inappropriate it would be to consider such a "priority queue" as a "queue" (same goes for stacks).

Furthermore, `scala.collection.mutable.PriorityQueue` does not inherit `mutable.Queue` whereas `java.util.PriorityQueue` does implement `java.util.Queue`.

**Comment by Jasper-M** [comment]
> Furthermore, scala.collection.mutable.PriorityQueue does not inherit mutable.Queue whereas java.util.PriorityQueue does implement java.util.Queue.

Indeed, so that difference in documentation actually makes sense.

**Comment by ruippeixotog** [comment]
I think that the best course of action is to document in an unambiguous way the guarantees given by each collection - even if it's to tell that the order is undefined, as with the current `PriorityQueue`. Given that there's no consensus of what the behavior of a theoretical priority queue implementation should be, I think it's better to leave things as they are.

Personally, I don't expect of a priority queue to guarantee dequeuing of same-priority elements in a FIFO manner, but I certainly understand the ambiguity. One question though, is the need for ordering stability big enough so that it would be useful to have a `StablePriorityQueue` implementation in the standard library?

**Comment by eyalroth** [comment]
@ruippeixotog I would argue that if the heap-like data structure -- which I believe is rarely being used -- is in the standard library, then the FIFO variant should be there as well.

Note at some of akka's contributors reactions to the issue from the two links I posted earlier:

patriknw:
> LGTM
> dimbleby great contribution. In my opinion this should replace the old priority mailbox. Undefined order for equal prio is confusing for actors. This will be less efficient, but if someone care about that they shouldn't use a prio mailbox in the first place. Let's see what others think.

viktorklang (on the SO question):
> Using priorities at all actually means that messages might be starved forever. OTOH I was also initially surprised. Looking forward to see your PriorityQueue implementation!

**Comment by diesalbla** [comment]
Given that the preferred solution was to document the existing implementation, which was done in https://github.com/scala/scala/pull/7144, and since there has been no further discussion, I think this issue is closed.
